### PR TITLE
GitLab and git-lfs enhance

### LINKED ISSUE #212: How to use object store for lfs
Gitlab now use minio object store, how to configure it?

### PR DESCRIPTION
**Pull request by icode**
* [GitLab] Support hashed store path
* Support remote git-lfs http server

close #212

### CODE REVIEW

File: src/main/java/svnserver/ext/gitlab/mapping/GitLabHookEvent.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
  */
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class GitLabHookEvent {
+  private static final ObjectMapper mapper = new ObjectMapper();
```
Comment (slonopotamus) [inline src/main/java/svnserver/ext/gitlab/mapping/GitLabHookEvent.java]: Is it thread-safe?

File: src/main/java/svnserver/server/SessionContext.java [hunk 2]
```diff
@@ -33,6 +33,8 @@
  * @author Artem V. Navrotskiy <bozaro@users.noreply.github.com>
  */
 public final class SessionContext {
+  static final ThreadLocal<SessionContext> LOCAL = new ThreadLocal<>();
```
Comment (slonopotamus) [inline src/main/java/svnserver/server/SessionContext.java]: Why you need this?

**Comment by slonopotamus** [comment]
Tests do not compile: https://travis-ci.org/bozaro/git-as-svn/jobs/453166346#L694

    /home/travis/build/bozaro/git-as-svn/src/test/java/svnserver/ext/gitlfs/storage/network/LfsHttpStorageTest.java:87: error: method getReader in class LfsHttpStorage cannot be applied to given types;

File: src/main/java/svnserver/ext/gitlab/mapping/GitLabHookEvent.java (hunk 1 above)
Comment (icode) [reply src/main/java/svnserver/ext/gitlab/mapping/GitLabHookEvent.java]: of course! 
> Mapper instances are fully thread-safe provided that ALL configuration of the instance occurs before ANY read or write calls. If configuration of a mapper is modified after first usage, changes may or may not take effect, and configuration calls themselves may fail. If you need to use different configuration, you have two main possibilities:
Construct and use ObjectReader for reading, ObjectWriter for writing. Both types are fully immutable and you can freely create new instances with different configuration using either factory methods of ObjectMapper, or readers/writers themselves. Construction of new ObjectReaders and ObjectWriters is a very light-weight operation so it is usually appropriate to create these on per-call basis, as needed, for configuring things like optional indentation of JSON.
If the specific kind of configurability is not available via ObjectReader and ObjectWriter, you may need to use multiple ObjectMapper instead (for example: you can not change mix-in annotations on-the-fly; or, set of custom (de)serializers). To help with this usage, you may want to use method copy() which creates a clone of the mapper with specific configuration, and allows configuration of the copied instance before it gets used. Note that copy() operation is as expensive as constructing a new ObjectMapper instance: if possible, you should still pool and reuse mappers if you intend to use them for multiple operations.

File: src/main/java/svnserver/server/SessionContext.java (hunk 2 above)
Comment (icode) [reply src/main/java/svnserver/server/SessionContext.java]: Get current socket user and repo info at any where in socket thread.
All lfs read interface not have `User` parameter, and repo info, url path mapper file path

**Comment by icode** [comment]
> Tests do not compile: https://travis-ci.org/bozaro/git-as-svn/jobs/453166346#L694
> 
> ```
> /home/travis/build/bozaro/git-as-svn/src/test/java/svnserver/ext/gitlfs/storage/network/LfsHttpStorageTest.java:87: error: method getReader in class LfsHttpStorage cannot be applied to given types;
> ```

I don't know this test unit why use not have interface, i want this interface `getReader(oid, user)`, but not, however test unit appeared， use this

**Comment by icode** [comment]
> Tests do not compile: https://travis-ci.org/bozaro/git-as-svn/jobs/453166346#L694
> 
> ```
> /home/travis/build/bozaro/git-as-svn/src/test/java/svnserver/ext/gitlfs/storage/network/LfsHttpStorageTest.java:87: error: method getReader in class LfsHttpStorage cannot be applied to given types;
> ```

I change it

**Comment by icode** [comment]
All checks have passed, please merge it!

**Comment by icode** [comment]
@slonopotamus please merge it!

**Comment by slonopotamus** [comment]
Hi, @icode. I'm currently working on a modified version of your PR. Most important change that I want to do is to remove threadLocal and instead access gitlab lfs not on behalf of connected user but on behalf of gitlab token that you specify in git-as-svn config. ETA: this or next week.

**Comment by slonopotamus** [comment]
Yeah, I do understand that. ThreadLocal approach means git-as-svn doesn't have objective picture of repository contents, instead he has a subjective view depending on current user. This is dangerous and can break various places where we cache stuff regardless on current user in the future.

You don't need to rebase your PR, I'll handle that.

**Comment by slonopotamus** [comment]
I'm not going to take this PR due to my concerns about thread local and the fact that we lose objective view on repository contents that can be dangerous to different caches git-as-svn uses. Instead, see #222.

**Pull request closed without merging**